### PR TITLE
Cmm filter test

### PR DESCRIPTION
**Pull request by chmorgan**

### CODE REVIEW

File: SharpPcap/LibPcap/PcapDevice.cs [hunk 1]
```diff
@@ -482,7 +482,8 @@ public virtual string Filter
         [EditorBrowsable(EditorBrowsableState.Advanced)]
         public static bool RunBpfProgram(IntPtr bpfProgram, IntPtr header, IntPtr data)
         {
-            return LibPcapSafeNativeMethods.pcap_offline_filter(bpfProgram, header, data);
+            var result = LibPcapSafeNativeMethods.pcap_offline_filter(bpfProgram, header, data);
+            return (result == 0) ? false : true;
```
Comment (kayoub5) [inline SharpPcap/LibPcap/PcapDevice.cs:486]:
```suggestion
            return result != 0;
```

File: Test/CheckFilterTest.cs [hunk 2]
```diff
@@ -18,5 +20,41 @@ public void TestFilters()
...
             Assert.IsNull(errorString);
         }
+
+        /// <summary>
+        /// Test RunBpfProgram() and other filter related methods
+        /// </summary>
+        [Test]
+        public void FilterMethods()
+        {
+            using var device = TestHelper.GetPcapDevice();
```
Comment (kayoub5) [inline Test/CheckFilterTest.cs:30]: @chmorgan this adapter is not necessarily the internet adapter, thus `WebFetch` may not have an effect on it.

Comment (chmorgan) [reply Test/CheckFilterTest.cs:30]: This same approach does work for the WinDivert test though. That's why I figured it would be ok here.

Comment (kayoub5) [reply Test/CheckFilterTest.cs:30]: Windivert is a layer 3 driver, libpcap is layer 2. a reliable solution would be to use the ARP class for example.

**Comment by chmorgan** [comment]
You mean use the arp class to generate traffic for the test? I can try that.

File: Test/CheckFilterTest.cs [hunk 3]
```diff
@@ -18,5 +20,40 @@ public void TestFilters()
...
+            device.Open();
+
+            var filterExpression = "tcp";
```
Comment (kayoub5) [inline Test/CheckFilterTest.cs:33]:
```suggestion
            var filterExpression = "arp";
```

File: Test/CheckFilterTest.cs [hunk 4]
```diff
@@ -18,5 +20,40 @@ public void TestFilters()
...
+            var header = IntPtr.Zero;
+            var data = IntPtr.Zero;
+            int retval = device.GetNextPacketPointers(ref header, ref data);
+
+            Assert.AreEqual(1, retval); // confirm we read a packet
+            Assert.AreNotEqual(IntPtr.Zero, header);
+            Assert.AreNotEqual(IntPtr.Zero, data);
+
+            // and test it against the bpf filter to confirm an exception is not thrown
+            Assert.DoesNotThrow(() => PcapDevice.RunBpfProgram(bpfProgram, header, data));
```
Comment (kayoub5) [inline Test/CheckFilterTest.cs:54]: Maybe check the return value?

File: SharpPcap/LibPcap/LibPcapSafeNativeMethods.Interop.cs [hunk 5]
```diff
@@ -126,7 +126,7 @@ internal static partial class LibPcapSafeNativeMethods
         /// </summary>
         /// <returns></returns>
         [DllImport(PCAP_DLL, CharSet = CharSet.Ansi, CallingConvention = CallingConvention.Cdecl)]
-        internal extern static bool pcap_offline_filter(IntPtr /*bpf_program **/ prog, IntPtr /* pcap_pkthdr* */ header, IntPtr pkt_data);
+        internal extern static int pcap_offline_filter(IntPtr /*bpf_program **/ prog, IntPtr /* pcap_pkthdr* */ header, IntPtr pkt_data);
```
Comment (kayoub5) [inline SharpPcap/LibPcap/LibPcapSafeNativeMethods.Interop.cs:129]: This looks like the kind of bug that would cause crashes similar to the WinDivert one, could we have it back-ported to v5.x  ?

File: Test/TestHelper.cs [hunk 6]
```diff
@@ -153,5 +154,16 @@ private static void Configuration_ConfigurationFailed(object sender, Configurati
         {
             Assert.Fail(e.Message);
         }
+
+        public static void WebFetch()
```
Comment (kayoub5) [inline Test/TestHelper.cs:158]: Is this change really necessary anymore?

File: SharpPcap/LibPcap/LibPcapSafeNativeMethods.Interop.cs (hunk 5 above)
Comment (chmorgan) [reply SharpPcap/LibPcap/LibPcapSafeNativeMethods.Interop.cs:129]: We could. My time to support the library has been high recently but we've never supported old major revs like we might if there were 10x the users or 2x the developers. And what happens if we release a 5.x release after releasing a 6.x release? So I'm not opposed if you want to figure out all of that stuff. It's just the process of making a 5.x branch at the point where we started on the 6.x changes, if we plan on actually supporting 5.x.

File: Test/TestHelper.cs (hunk 6 above)
Comment (chmorgan) [reply Test/TestHelper.cs:158]: It isn't. We could drop it if we don't expect other tests to need to generate web requests to generate traffic.

Comment (kayoub5) [reply Test/TestHelper.cs:158]: Not likely

File: SharpPcap/LibPcap/LibPcapSafeNativeMethods.Interop.cs (hunk 5 above)
Comment (kayoub5) [reply SharpPcap/LibPcap/LibPcapSafeNativeMethods.Interop.cs:129]: the v6.x release is not going to be soon, patching this for v5.x would be nice while we finalize v6.x

**Review by kayoub5 — APPROVED** [review]
